python/xir/program.py

```diff
@@ -41,46 +43,83 @@ class Statement:
         name (str): name of the statement
         params (list, Dict): parameters for the statement (can be empty)
         wires (tuple): the wires on which the statement is applied
+
+    Keyword args:
+        adjoint (bool): whether the statement is an adjoint gate
+        ctrl_wires (tuple): the control wires of a controlled gate statement
         use_floats (bool): Whether floats and complex types are returned instead of ``Decimal``
             and ``DecimalComplex`` objects. Defaults to ``True``.
     """
 
-    def __init__(self, name: str, params: Union[List, Dict], wires: Tuple, use_floats: bool = True):
+    def __init__(
+            self,
+            name: str,
+            params: Union[List, Dict],
+            wires: Tuple,
+            **kwargs
+        ):
         self._name = name
         self._params = params
         self._wires = wires
 
-        self._use_floats = use_floats
+        self._is_adjoint = kwargs.get("adjoint", False)
+        self._ctrl_wires = kwargs.get("ctrl_wires", tuple())
+
+        self._use_floats = kwargs.get("use_floats", True)
 
     def __str__(self):
+        """Serialized string representation of a Statement"""
         if isinstance(self.params, dict):
             params = [f"{k}: {v}" for k, v in self.params.items()]
         else:
             params = [str(p) for p in self.params]
         params_str = ", ".join(params)
+        if params_str != "":
+            params_str = "(" + params_str + ")"
 
         wires = ", ".join([str(w) for w in self.wires])
 
-        if params_str == "":
-            return f"{self.name} | [{wires}]"
-        return f"{self.name}({params_str}) | [{wires}]"
+        modifier_str = ""
+        if len(self.ctrl_wires) != 0:
+            ctrl_wires = ", ".join([str(w) for w in self.ctrl_wires])
+            modifier_str = f"ctrl[{ctrl_wires}] "
+        if self.is_adjoint:
+            modifier_str = "adjoint " + modifier_str
+
+        return f"{modifier_str}{self.name}{params_str} | [{wires}]"
 
     @property
     def name(self) -> str:
+        """Returns the name of the gate statement"""
         return self._name
 
     @property
     def params(self) -> Union[List, Dict]:
+        """Returns the parameters of the gate statement"""
         if self.use_floats:
             return get_floats(self._params)
         return self._params
 
     @property
-    def wires(self) -> Tuple:
+    def wires(self) -> Tuple[Wire, ...]:
+        """Returns the wires that the gate is applied to"""
         return self._wires
 
+    @property
+    def is_adjoint(self) -> bool:
+        """Whether the statement is an adjoint gate"""
+        return self._is_adjoint
+
+    @property
+    def ctrl_wires(self) -> Tuple[Wire, ...]:
+        """Returns the control wires of a controlled gate statement"""
```

Comment:
I think it is worth documenting the return value when no control wires are specified:
```suggestion
        """Returns the control wires of a controlled gate statement.
        If no control wires are specified, an empty tuple is returned.
        """
```